src/componentSetBuilder.ts

```diff
@@ -38,64 +37,69 @@ export class ComponentSetBuilder {
    */
   public static async build(options: ComponentSetOptions): Promise<ComponentSet> {
     const logger = Logger.childFromRoot('createComponentSet');
-    const csAggregator: ComponentLike[] = [];
+    let componentSet: ComponentSet;
 
     const { sourcepath, manifest, metadata, packagenames, apiversion, sourceapiversion } = options;
     try {
       if (sourcepath) {
         logger.debug(`Building ComponentSet from sourcepath: ${sourcepath.toString()}`);
+        const fsPaths: string[] = [];
         sourcepath.forEach((filepath) => {
-          if (fs.fileExistsSync(filepath)) {
-            csAggregator.push(...ComponentSet.fromSource(path.resolve(filepath)));
-          } else {
+          if (!fs.fileExistsSync(filepath)) {
             throw new SfdxError(`The sourcepath "${filepath}" is not a valid source file path.`);
           }
+          fsPaths.push(path.resolve(filepath));
         });
+        componentSet = ComponentSet.fromSource({ fsPaths });
       }
 
       // Return empty ComponentSet and use packageNames in the library via `.retrieve` options
       if (packagenames) {
         logger.debug(`Building ComponentSet for packagenames: ${packagenames.toString()}`);
-        csAggregator.push(...new ComponentSet([]));
+        componentSet ??= new ComponentSet();
       }
 
       // Resolve manifest with source in package directories.
       if (manifest) {
         logger.debug(`Building ComponentSet from manifest: ${manifest.manifestPath}`);
         const directoryPaths = options.manifest.directoryPaths;
         logger.debug(`Searching in packageDir: ${directoryPaths.join(', ')} for matching metadata`);
-        const compSet = await ComponentSet.fromManifest({
+        componentSet = await ComponentSet.fromManifest({
           manifestPath: manifest.manifestPath,
           resolveSourcePaths: options.manifest.directoryPaths,
           forceAddWildcards: true,
         });
-        csAggregator.push(...compSet);
       }
 
       // Resolve metadata entries with source in package directories.
       if (metadata) {
         logger.debug(`Building ComponentSet from metadata: ${metadata.metadataEntries.toString()}`);
         const registry = new RegistryAccess();
+        const compSetFilter = new ComponentSet();
+        componentSet ??= new ComponentSet();
```

Comment:
Trying to figure out the intent of `??=`  Is there a possibility that you could have more than one of  [metadata, sourcepath, packagename, manifest] where componentSet would already be defined?

If they are mutually exclusive, we could probably provide a more expressive type to `ComponentSetOptions` to prevent users from passing in multiple?

Reply:
`source:retrieve --help` output shows `sfdx force:source:retrieve -n MyPackageName -p path/to/apex/classes` and the other variations of `-n`, `-x` and `-m`. These options aren't mutually exclusive